homeassistant/components/climate/__init__.py

```diff
@@ -556,6 +558,19 @@ async def async_turn_off(self) -> None:
         if HVACMode.OFF in self.hvac_modes:
             await self.async_set_hvac_mode(HVACMode.OFF)
 
+    async def async_toggle(self) -> None:
+        """Toggle the entity."""
+        if hasattr(self, "toggle"):
+            await self.hass.async_add_executor_job(self.toggle)
+            return
+
+        # If OFF is in the HVAC modes, then we can use it to decide what to do.
+        if HVACMode.OFF in self.hvac_modes:
+            if self.hvac_mode == HVACMode.OFF:
+                await self.async_turn_on()
+            else:
+                await self.async_turn_off()
```

Comment:
What do we do if OFF is NOT in the HVAC modes?

The existing `async_turn_off` does nothing in that case, but I think we should add a `ClimateEntityFeature.TURN_OFF` flag which is automatically set if either `HVACMode.OFF in self.hvac_modes` or either of `async_toggle` or `toggle` are overridden / implemented.

Reply:
@emontnemery should we perhaps make a preliminary PR to implement it for the existing `turn_off` before adding on this?

Reply:
I'm not opposed to extending the turn_off functionality, but I do think it deserves its own PR.
This PR is intentionally narrow, its focus is to just provide more consistency, where if the climate has both turn_on and turn_off, it follows that it should have toggle as well.
Making turn_off function better/cover more cases is in my view orthogonal to this, and can be done later if the decision to pursue it is made.

Reply:
> What do we do if OFF is NOT in the HVAC modes?

We raise and fail the service.

Reply:
Agreed it should have it's own architecture issue + PR 👍 

Architecture issue: https://github.com/home-assistant/architecture/discussions/982
PR: https://github.com/home-assistant/core/pull/101673